Utilities/DataLoading/EdFi.LoadTools/Engine/ResourcePipeline/SubmitResourceTimingDecorator.cs

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using log4net;
+
+namespace EdFi.LoadTools.Engine.ResourcePipeline
+{
+    public class SubmitResourceTimingDecorator : ISubmitResource
+    {
+        private readonly ISubmitResource _next;
+        private readonly ResourceStatistic _resourceStatistic;
+        private readonly ILog _log = LogManager.GetLogger(nameof(ApiLoaderApplication));
+        private int _count;
+
+        public SubmitResourceTimingDecorator(ISubmitResource next, ResourceStatistic resourceStatistic)
+        {
+            _next = next;
+            _resourceStatistic = resourceStatistic;
+        }
+
+        public async Task<ApiLoaderWorkItem> ProcessAsync(ApiLoaderWorkItem resourceWorkItem)
+        {
+            var contextPrefix = LogContext.BuildContextPrefix(resourceWorkItem);
+            var sw = Stopwatch.StartNew();
+            var count = Interlocked.Increment(ref _count);
+
+            var results = await _next.ProcessAsync(resourceWorkItem);
+            sw.Stop();
+
+            _log.Debug(
+                $"{contextPrefix} #{count} completed in {Math.Round((decimal) sw.Elapsed.TotalMilliseconds, 3)} milliseconds");
+
+            _resourceStatistic.Resources++;
```

Comment:
Looking at this more closely, one of the goals I was hoping to get was the ability to see the average time on a per resource level.  Averaging across the entire load isn't necessarily helpful, since there can be a pretty wide disparity between response times, which just gets flattened out over large numbers.  With the structure you have it should be pretty easy to let AddResourceTiming take in the ResourceWorkItem information, and then internally to the structure store the information about number of resources and milliseconds on a per resource basis.  Then in the final report out, you can show information on total resources and time per resource per resource, and an overall, with no additional data collection.